userscripts/view_in_mpv: using yt-dlp to download video

sim590 opened the pull request:
This makes `view_in_mpv` userscript use `yt-dlp` for download.

As mentioned [here](https://github.com/mpv-player/mpv/issues/8655#issuecomment-1199594426), mpv doesn't use `yt-dlp` for downloading, but only gets an URL back from it and lets `ffmpeg` download the video. This is very slow as ffmpeg hasn't implemented same optimizations that yt-dlp has for making youtube downloads quick.

In order to go around that, it's suggested to use `yt-dlp` and `mpv` this way:

```sh
yt-dlp -o - $URL | mpv $FLAGS -
```

This may cause issues when seeking with `mpv`  stated [here](https://github.com/mpv-player/mpv/issues/8655#issuecomment-1199733728), but I can't confirm as downloads are so quick on my computer with this that I didn't see any issue yet.

For instance, trying to watch the following in MPV with the original userscript:

https://www.youtube.com/watch?v=UhM_H3lFk_Q

was sow slow that, buffering was stopped the video every 1-2 seconds. Enabling `yt-dlp` made the download instantaneous.

There is an environment variable (`USE_YTDLP_FOR_DOWNLOAD`) for enabling download through `yt-dlp` in order to keep compatibility with past version and since there's an additional dependency.

sim590 commented:
I just thought about the fact that the title of the MPV window won't have the title of the video in the window title...

![image](https://user-images.githubusercontent.com/4147254/218246483-a09e6dc1-2e7b-4338-aae9-8f6c26e4117f.png)

I will try to find a way to fix this.

sim590 commented:
> I just thought about the fact that the title of the MPV window won't have the title of the video in the window title...
> 
> ![image](https://user-images.githubusercontent.com/4147254/218246483-a09e6dc1-2e7b-4338-aae9-8f6c26e4117f.png)
> 
> I will try to find a way to fix this.

I tested the following:
```bash
video_title=$(yt-dlp --get-title "$QUTE_URL" 2>/dev/null | head -n1)
tmpdir=$(mktemp -d)
yt-dlp "$YT_DLP_FLAGS" -o - "$QUTE_URL" >"$tmpdir/$video_title" 2>/dev/null &
yt_dlp_pid=$!
sleep 2
mpv --no-terminal "$tmpdir/$video_title"
ps -p $yt_dlp_pid && kill -9 $yt_dlp_pid
rm -rf $tmpdir
```
and it works. But I don't like the `sleep 2` instruction. I'd like a better suggestion before I push this (if it's even desirable solution).

The-Compiler commented:
hm, seems a bit overkill to me - I'd prefer a simpler solution.

At the cost of a second invocation taking a second or two, couldn't you do something like `... | mpv --title "$(yt-dlp --get-title "$QUTE_URL")"` instead?

sim590 commented:
Wow. I didn't find `--title`. That's awesome! :D

The-Compiler commented:
Seems fine to me. @t-wissmann any comments from your side?

sim590 commented:
I've actually just stumbled on the downside of this method and it manifests itself in that you cannot seek backwards at some point.

I'm not sure, but may be the following MPV flags `--demuxer-max-bytes` and/or `--demuxer-max-back-bytes` could help? Like setting: `--demuxer-max-bytes=1G --demuxer-max-back-bytes=1G` ? That would depend on the video size.

t-wissmann commented:
> Seems fine to me. @t-wissmann any comments from your side?

Looks good to me. But I haven't used `yt-dlp` yet

sim590 commented:
I fixed the parsing of `YT_DLP_FLAGS` which was not right after I tested it.

I also realized a disappointing thing when using this method: as soon as a format is specified for `yt-dlp`, then the downloader switches to FFMPEG downloader which reverts the situation back to the initial issue. Using `--downloader=http` in `yt-dlp` however leaves us with no audio. The workaround I found was to download audio separately.

I suggest the user to wrap the `view_in_mpv` script with something along the following lines:

```sh
#!/usr/bin/env bash

PATH_TO_QUTEB_MPV_SCRIPT=$HOME/prog/qutebrowser/misc/userscripts/view_in_mpv

tmpdir=$(mktemp -d)
yt-dlp -f 'bestaudio' -o - "$QUTE_URL" >$tmpdir/ytdlp_audio
env USE_YTDLP_FOR_DOWNLOAD=yes \
    YT_DLP_FLAGS="--downloader=http --format=bestvideo[height<=1440]" \
    MPV_FLAGS="--audio-file=$tmpdir/ytdlp_audio --force-window --no-terminal --keep-open=yes --demuxer-max-bytes=2G" \
    $PATH_TO_QUTEB_MPV_SCRIPT
rm -rf $tmpdir
```

This enables downloading videos at great speed while keeping audio, of course. I'm not sure if that could have been simpler, but ... This is somewhat similar to what is suggested by others [here](https://github.com/mpv-player/mpv/issues/8655#issuecomment-1279012262).

toofar commented:
Looks reasonable to me, it's sure a lot faster. As you point out, there are still a few corner cases and places where the user experience isn't as good. But if it's opt-in it seems fine. What do you think about adding an argument to control the behaviour instead of just the env var? Eg

```diff
--- a/misc/userscripts/view_in_mpv        2023-08-20 11:32:48.727569354 +1200
+++ b/misc/userscripts/view_in_mpv      2023-08-20 11:32:10.560521544 +1200
@@ -1,16 +1,22 @@
 #!/usr/bin/env bash
 #
 # Behavior:
-#   Userscript for qutebrowser which views the current web page in mpv using
+#   Userscript for qutebrowser which views the current youtube video in mpv using
 #   sensible mpv-flags. While viewing the page in MPV, all <video>, <embed>,
 #   and <object> tags in the original page are temporarily removed. Clicking on
 #   such a removed video restores the respective video.
 #
 #   In order to use this script, just start it using `spawn --userscript` from
-#   qutebrowser. I recommend using an alias, e.g. put this in the
-#   [alias]-section of qutebrowser.conf:
+#   qutebrowser. I recommend using an alias, e.g. put this in your config.py:
 #
-#     mpv = spawn --userscript /path/to/view_in_mpv
+#     c.aliases["mpv"] = "spawn --userscript /path/to/view_in_mpv"
+#
+# Arguments:
+#   * `-f`: Fast download, spawn yt-dlp to download the video, instead of
+#     letting mpv do it. This allows for much faster video downloads but is a
+#     bit more fragile: a) if you seek past the end of the mpv cache it'll
+#     likely break b) if you have custom formats set for yt-dlp the download
+#     speed will regress again.
 #
 # Background:
 #   Most of my machines are too slow to play youtube videos using html5, but
@@ -50,10 +56,21 @@
 MPV_COMMAND=${MPV_COMMAND:-mpv}
 YTDLP_COMMAND=${YTDLP_COMMAND:-yt-dlp}
 # Warning: spaces in single flags are not supported
 MPV_FLAGS=${MPV_FLAGS:- --force-window --no-terminal --keep-open=yes --ytdl}
 IFS=" " read -r -a video_command <<< "$MPV_COMMAND $MPV_FLAGS"
 IFS=" " read -r -a yt_dlp_video_command <<< "$YTDLP_COMMAND $YT_DLP_FLAGS"

+case "$1" in
+  "-f")
+    USE_YTDLP_FOR_DOWNLOAD="yes"
+    shift
+    ;;
+  *)
+    msg error "$0: Unknown arg '$1'"
+    shift
+    ;;
+esac
+
 js() {
 cat <<EOF

```

sim590 commented:
By the way, the JS script doesn't work anymore on Youtube. It has been not working for a while now. I don't know how it works, but it would be cool to fix it someday.

sim590 commented:
I'll mention that I also changed some stuff:

* I changed the location where the audio files are saved. I create the directory here `$HOME/.cache/qutebrowser/view_in_mpv` and create a temporary one for each audio download. Could have been just a random file name under `view_in_mpv` instead of a random directory containing the audio file. Tell me if you want me to change it.
* I added a `-u` flag in order to allow to use the script with `{hint-url}` in your key maps. In that case, I also disable the javascript script.
* I added a flag to `yt-dlp` calls in order to allow using the Qutebrowser cookies to download. In some cases, it's required to avoid errors.

sim590 commented:
**Important update: the speed issue has been fixed upstream**

The slow download speed that motivated this PR has been fixed directly in FFmpeg and mpv:

- **FFmpeg**: [PR #21703](https://code.ffmpeg.org/FFmpeg/FFmpeg/pulls/21703) adds a `request_size` AVOption to the HTTP protocol, allowing downloads to be split into small range requests. YouTube no longer throttles downloads with this approach.
- **mpv**: [PR #17471](https://github.com/mpv-player/mpv/pull/17471) (commit `84229f9`, merged Feb 27, 2026) updates `ytdl_hook.lua` to automatically read the `http_chunk_size` key from yt-dlp's JSON output and pass it as `request_size` to FFmpeg.

With mpv post-v0.41.0 and FFmpeg >= n8.1, `mpv <URL>` downloads at full speed **and** seeking works normally — no need for `yt-dlp -o - | mpv -` or separate audio downloading.

For users on mpv v0.41.0 with FFmpeg >= n8.1, the workaround `stream-lavf-o-append=request_size=10485760` in `mpv.conf` is sufficient.

I will close my PR as I think it is useless now.
